### PR TITLE
#16431 resolve driver properties

### LINKED ISSUE #16431: use variables in connection's driver properties won't work
#### System information: 
- Ubuntu20.04
- DBeaver version 22.04

#### Connection specification:
- Mariadb
- mariadb jdbc
- socks 

#### Describe the problem you're observing:
If I use full path in the driver properties it works, but neither ${home} nor ${HOME} work as path.

### PR DESCRIPTION
**Pull request by Destrolaric**
Closes #16431 
Has limitations:
1. Resolve will happen once, after that resolved value will be used
2. Resolving happens only in connection configuration stage

### CODE REVIEW

**Review by E1izabeth — APPROVED** [review]

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/properties/PropertySourceCustom.java [hunk 1]
```diff
@@ -98,6 +99,8 @@ public void addDefaultValues(Map<String, ?> defaultValues)
     public Map<String, Object> getPropertyValues() {
         Map<String, Object> allValues = new HashMap<>(originalValues);
         allValues.putAll(propValues);
+        allValues.values().forEach(it -> GeneralUtils.replaceVariables(CommonUtils.toString(it),
+            SystemVariablesResolver.INSTANCE));
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/properties/PropertySourceCustom.java]: I'm suspicious about converting property values from * something * into string just to expand variables.

If variable values can only be strings, then it's better to return `Map<String, String>` instead.

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/connection/DriverPropertiesDialogPage.java [hunk 2]
```diff
@@ -123,7 +127,8 @@ public void saveSettings(DBPDataSourceContainer dataSource) {
             for (Map.Entry<String, Object> entry : propertySource.getPropertyValues().entrySet()) {
                 String propName = CommonUtils.toString(entry.getKey());
                 if (!propName.isEmpty()) {
-                    properties.put(propName, CommonUtils.toString(entry.getValue()));
+                    properties.put(propName, GeneralUtils.replaceVariables(CommonUtils.toString(entry.getValue()),
+                        new SystemVariablesResolver()));
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/connection/DriverPropertiesDialogPage.java:131]: See previous comment

File: plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/connection/DriverPropertiesDialogPage.java [hunk 3]
```diff
@@ -174,7 +179,11 @@ public void widgetSelected(SelectionEvent e) {
             }
             netConfigLink.setLayoutData(new GridData(GridData.HORIZONTAL_ALIGN_END));
         }
-
+        VariablesHintLabel variablesHintLabel = new VariablesHintLabel(ph,
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ui.editors.connection/src/org/jkiss/dbeaver/ui/dialogs/connection/DriverPropertiesDialogPage.java:182]: Redundant variable declaration

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

**Comment by Matvey16** [comment]
Verified

**Pull request closed without merging**